pkg/network/server.go

```diff
@@ -331,12 +332,30 @@ func (s *Server) startProtocol(p Peer) {
 			err = p.WriteMsg(m)
 		case <-timer.C:
 			// Try to sync in headers and block with the peer if his block height is higher then ours.
-			if p.Version().StartHeight > s.chain.BlockHeight() {
+			if p.LastBlockIndex() > s.chain.BlockHeight() {
 				err = s.requestBlocks(p)
 			}
 			if err == nil {
 				timer.Reset(s.ProtoTickInterval)
 			}
+			if s.chain.HeaderHeight() >= p.LastBlockIndex() {
+				block, errGetBlock := s.chain.GetBlock(s.chain.CurrentBlockHash())
+				if errGetBlock != nil {
+					err = errGetBlock
+				} else {
+					diff := uint32(time.Now().UTC().Unix()) - block.Timestamp
+					if diff > uint32(s.PingInterval/time.Second) {
+						p.UpdatePingSent(true)
+						err = p.WriteMsg(NewMessage(s.Net, CMDPing, payload.NewPing(s.id, s.chain.HeaderHeight())))
+					}
+				}
+			}
+		case <-pingTimer.C:
+			if p.PingSent() {
```

Comment:
What do you think about counting some number of pings before disconnect, say three for example? It might be too aggressive disconnecting immediately after the first failed ping attempt.

Reply:
yup, why not to count till 3 =)